server/controllers/stock/index.js

```diff
@@ -129,78 +130,205 @@ function createStock(req, res, next) {
+      description : params.movement.description,
+      period_id : periodId,
+    });
+  });
 
-        // adding a movement insertion query into the transaction
-        transaction.addQuery(`INSERT INTO stock_movement SET ?`, {
-          uuid : db.bid(uuid()),
-          lot_uuid : db.bid(lotUuid),
-          depot_uuid : db.bid(params.movement.depot_uuid),
-          document_uuid : db.bid(documentUuid),
-          flux_id : params.movement.flux_id,
-          date : new Date(params.movement.date),
-          quantity : lot.quantity,
-          unit_cost : lot.unit_cost,
-          is_exit : 0,
-          user_id : params.movement.user_id,
-          description : params.movement.description,
-          period_id : periodId,
-        });
-      });
+  const postingParams = [
+    db.bid(documentUuid), 0, session.project.id, session.enterprise.currency_id,
+  ];
 
-      const postingParams = [db.bid(documentUuid), 0, req.session.project.id, req.session.enterprise.currency_id];
+  if (session.enterprise.settings.enable_auto_stock_accounting) {
+    transaction.addQuery('CALL PostStockMovement(?)', [postingParams]);
+  }
 
-      if (req.session.enterprise.settings.enable_auto_stock_accounting) {
-        transaction.addQuery('CALL PostStockMovement(?)', [postingParams]);
-      }
+  await transaction.execute();
+  return documentUuid;
+}
 
-      // execute all operations as one transaction
-      return transaction.execute();
-    })
-    .then(() => {
+/**
+ * POST /stock/integration
+ * create a new integration entry
+ */
+function createIntegration(req, res, next) {
+  insertNewStock(req.session, req.body, 'integration')
+    .then(documentUuid => {
       res.status(201).json({ uuid : documentUuid });
     })
-    .catch(next)
-    .done();
+    .catch(next);
+}
+
+/**
+ * POST /stock/inventory_adjustment
+ * stock inventory adjustement
+ */
+async function createInventoryAdjustment(req, res, next) {
+  try {
+    const movement = req.body;
+
+    const lots = movement.lots
+      .filter(l => l.quantity !== l.oldQuantity);
+
+    const period = await Fiscal.lookupFiscalYearByDate(new Date(movement.date));
+    const periodId = period.id;
+
+    if (!movement.depot_uuid) {
+      throw new Error('No defined depot');
+    }
+
+    // selected lots identifiers
+    const inventoryLotUuids = lots.map(l => l.uuid);
+
+    // get list of lots with their quantities
+    const allLots = await core.getLotsDepot(null, { depot_uuid : movement.depot_uuid });
+
+    // reverse all selected lots
+    const availableLots = allLots.filter(lot => {
+      return inventoryLotUuids.includes(lot.uuid);
+    });
+
+    // pass reverse operations
+    const trx = db.transaction();
+
+    const positiveAdjustmentUuid = uuid();
+    const negativeAdjustmentUuid = uuid();
+    const positiveQuantities = availableLots.filter(lot => lot.quantity > 0);
+    const negativeQuantities = availableLots.filter(lot => lot.quantity < 0);
```

Comment:
I can't figure out where we will get negative quantities from.  Where do they come from?

Reply:
We had found some stock with negative quantities at IMCK in the stock because of some early stock module which allowed to escape validation of stock quantity during the stock exit process

Reply:
Aha!  That makes much more sense.  Can you leave a comment saying that?  I was combing through your code trying to figure out were this would have been possible.